feat(ui): add autosuggest for seed recovery

bolatovumar opened the pull request:
## Description:

Add autosuggest functionality when on the seed recovery screen. Words are shown in a list below the input as user starts typing the word. Words can be selected from the suggestion list either by clicking on them or through the keyboard.

## Motivation and Context:

Address #2765

## How Has This Been Tested?

Manually

## Types of changes:

New feature: autosuggest seed words from BIP 39 when recovering your seed.

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have reviewed and updated the documentation accordingly.
- [x] I have read the _CONTRIBUTING_ document.
- [x] I have added tests to cover my changes where needed.
- [x] All new and existing tests passed.
- [x] My commits have been squashed into a concise set of changes.

korhaliv commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -40,32 +154,94 @@ class SeedWord extends React.Component {
   withFormApi,
-  injectIntl
+  injectIntl,
+  withTheme
```
This is probably used just for the bg here 
https://github.com/LN-Zap/zap-desktop/pull/2896/files#diff-378f70cc56cf8c3617c6784bc0c1dbb1R70
and should go once you refactor that

bolatovumar replied:
Yeah, will do.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -1,26 +1,134 @@
+    return null
+  }
+
+  const wordList = wordsToFilter.filter(item => item.indexOf(inputValue) === 0)
```
`item.indexOf(inputValue) < 0` won't work as that would find items that do not match the input value. `.indexOf(inputValue) === 0` matches the input value string exactly from the beginning. So, for example, if a user input `app` the word `apple` would match since `app` starts at the zero index of `apple`.

I will work on this to make it more obvious what's happening here.

bolatovumar replied:
Tried to make it more clear. Should be better now.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -1,26 +1,134 @@
+          as="li"
+          bg={highlightedIndex === index ? theme.colors.primaryColor : null}
+          sx={{
+            p: 2,
```
Noted. Will do.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -1,26 +1,134 @@
+            cursor: 'pointer',
+          }}
+        >
+          <Flex alignItems="center" pr={2}>
```
Probably not. Will see if I can remove `Box`.

bolatovumar replied:
Removed `Box`.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -1,26 +1,134 @@
+    }
+  }
+
+  componentDidUpdate(prevProps) {
```
I agree that it's probably not best to have `componentDidUpdate` but not sure how else to handle the case where a use paste in the entire 24 word seed and we need to update each input with new values.

How about I convert this component to a functional one with hooks in a follow-up PR?

bolatovumar replied:
Actually `componentDidUpdate` is not actually needed at all if `state` variables are removed.

bolatovumar replied:
Removed this code and made the component a functional component.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -82,6 +258,30 @@ class Recover extends React.Component {
+   * we don't have to filter through the whole array when looking up values
+   * every time
+   */
+  bip39WordMap = bip39.reduce((bip39WordMap, word) => {
```
Yeah, makes sense. Will do.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -1,26 +1,134 @@
+  box-shadow: ${themeGet('shadows.s')};
+`
+
+const SeedWordList = props => {
```
Will do.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -82,6 +258,30 @@ class Recover extends React.Component {
+   * every time
+   */
+  bip39WordMap = bip39.reduce((bip39WordMap, word) => {
+    var firstLetter = word[0]
```
Makes sense. Will do.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -131,40 +345,42 @@ class Recover extends React.Component {
+                      index={index + 1}
+                      onPaste={index === 0 ? this.handlePaste : null}
+                      setFormApiFieldValue={word => {
+                        this.setFormApiFieldValue(index)(word)
```
I did so that the partially applied function can be called like this `setFormApiFieldValue(selection)` in `SeedWord`. I guess it's not really needed since we pass the index to `SeedWord` and can just call the function like this `setFormApiFieldValue(index - 1, selection)` but it doesn't look as nice?

What do you think? Remove curried function?

bolatovumar replied:
Kept the curry for now unless we really wanna remove it.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -1,26 +1,134 @@
+            item,
+          })}
+          as="li"
+          bg={highlightedIndex === index ? theme.colors.primaryColor : null}
```
Thanks. Will do.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -82,6 +258,30 @@ class Recover extends React.Component {
+    return bip39WordMap
+  }, {})
+
+  constructor(props) {
```
@korhaliv Looking at this I realized there is no need to pass the seed as a prop to `Recover` component. This seed is never used for anything useful since you can't type out part of the seed, proceed further and return back to the recover form to see your partially typed out seed. You can click on the `Start over` button but this starts the process all over again with an empty seed.

Is it OK to remove this prop completely then?

bolatovumar replied:
Also, you are right about the `state` variable. None of the state variables are actually needed and just `formState` is sufficient. I will remove them. If there is a good reason to have these state variables they can always be brought back.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -40,32 +154,94 @@ class SeedWord extends React.Component {
+                  value,
+                  variant: 'thin',
+                  willAutoFocus: index === 1,
+                  fieldState: {
```
I thought I needed this but now I removed it and nothing breaks or changes at all as far as I can tell. If it turns out this is really needed it can always come back.

bolatovumar replied:
Done.

bolatovumar commented on renderer/components/Onboarding/Steps/Recover.js:
```diff
@@ -131,40 +345,42 @@ class Recover extends React.Component {
+                    <SeedWordWithIntl
+                      key={index + 1}
+                      bip39WordMap={this.bip39WordMap}
+                      formState={formState}
```
Done.

bolatovumar commented:
One thing I noticed when making this PR is that the screen sometimes flickers when you type out your seed. I first thought it was because of the new code I added but then I noticed this screen flicker exists on the `next` branch as well (without this new code) so it's not this PR that introduced it.

I'm not sure if this is just a one-off issue that only I'm experiencing or if others also experience it. If you do, it should probably be raised as a separate issue.

mrfelton commented on renderer/components/Onboarding/Steps/SeedWordList.js:
```diff
@@ -0,0 +1,103 @@
+  return bip39WordMap
+}, {})
+
+const StyledOptionList = styled.ul`
```
Generally speaking, we prefer to use rebass over raw styled-components as this generally results in more flexible components that support all of the features of rebass.

This could be re-written something more like this:

```jsx
const StyledOptionList = React.forwardRef(({ ...rest }, ref) => (
  <Box
    ref={ref}
    as="ul"
    bg="secondaryColor"
    maxHeight={300}
    mt={1}
    p={0}
    sx={{
      position: 'absolute',
      zIndex: 2,
      overflowY: 'auto',
      overflowX: 'hidden',
      outline: 0,
      transition: 'opacity 0.1s ease',
      borderRadius: 's',
      boxShadow: 's',
    }}
    width={1}
    {...rest}
  />
))
StyledOptionList.displayName = 'StyledOptionList'
```

bolatovumar replied:
Done.

bolatovumar commented:
> Nice one @bolatovumar .
> 
> A couple of suggestions on the code.
> 
> Also, I get the following warning in the console when typing into one of the seed word input fields
> 
> > backend.js:1 Warning: A component is changing an uncontrolled input of type undefined to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a

Yeah, I noticed this as well. I believe this is the result of me removing all of the local state from these components. The issue here is that I initially pass the value to the input as a prop from the parent component. This is needed for the case when a used copy-pastes the seed. At this point it's a controlled input. As soon as a user starts typing, it becomes an uncontrolled component. This could be resolved in two ways:
1. Introduce local state to the input component. We probably don't really wanna do this as local state is superfluous here?
2. Don't pass the value as a prop to the input. Need to ensure that copy-pasting the seed still works in this case.

I guess we could also ignore the warning as nothing breaks really. 
I will see if option 2 works and will report back.

bolatovumar commented on renderer/components/Onboarding/Steps/SeedWordInput.js:
```diff
@@ -0,0 +1,90 @@
+import React from 'react'
```
Done.

bolatovumar commented:
I tried not passing the `word` prop to set the input value to avoid the console error and just used `formApi.setValue` but in this case text doesn't show up in the input. Not sure what to do here. Use a ref, perhaps, and set the value directly on the underlying DOM element? What do you think @mrfelton?

mrfelton commented:
> I tried not passing the `word` prop to set the input value to avoid the console error and just used `formApi.setValue` but in this case text doesn't show up in the input. Not sure what to do here. Use a ref, perhaps, and set the value directly on the underlying DOM element? What do you think @mrfelton?

@bolatovumar see my commit here https://github.com/LN-Zap/zap-desktop/pull/2966/commits/122fa74b52e8aa457f24c020dc2c9ba768704a8d which I think addresses this. What do you think?

bolatovumar commented:
@mrfelton, yup that workaround seems to work. I will update my commit to include this workaround.

mrfelton commented:
@bolatovumar there are a number of other fixes and updates that have also been applied in 2966.

See conversation in #2966 and `git diff bolatovumar/feat-2765 mrfelton/feat-2765` for details.

As well as the fixes mentioned in the comments, I also implemented a solution to ensure that the dropdown height does not exceed the remaining height of the page so that items don't drop off the bottom of the screen (see https://github.com/LN-Zap/zap-desktop/pull/2966/commits/a2cec0923bbd0e300ae4e231fe743cf17ee7d611)

Also, fixed some issues with the tests (See https://github.com/LN-Zap/zap-desktop/pull/2966/commits/7a0d84dbe8fbe91eaaeb2e28beb256b85dbe1a89)

I think the only remaining issue was this one, reported by @korhaliv in slack:

![image](https://user-images.githubusercontent.com/200251/66697956-4a97d180-ecda-11e9-8894-8a96efe5f3fb.png)

> If field is invalid it won't show me suggestions popup i need to first remove the contents.

If you want to address that issue,  please could you rebase your branch off my one so that the other fixes are incorporated please? Thanks

bolatovumar commented:
@mrfelton I have merged your feature branch into mine, squashed the commits into one and rebased on top of the `next` branch.

Regarding this:
>If field is invalid it won't show me suggestions popup i need to first remove the contents.

I either don't understand what the issue is or it's not an issue. I recorded a video trying to replicate this issue but couldn't. See here:
https://youtu.be/5qBQzIySH_0

Let me know if I misunderstand something.

mrfelton commented:
@bolatovumar I've merged this via https://github.com/LN-Zap/zap-desktop/pull/2966 

We prefer to keep the commit authorship and granularity where possible rather than squashing everything up into a single commit (as long as each individual commit makes sense in it's own right).

Thanks for your work in this - it's a great feature addition
